src/app/clusters/operational-state-server/operational-state-server.cpp

```diff
@@ -0,0 +1,413 @@
+}
+
+CHIP_ERROR OperationalStateServer::SetOperationalState(OperationalStateStruct & opState)
+{
+    OperationalStateStructDynamicList op;
+
+    op.operationalStateID = static_cast<Clusters::OperationalState::OperationalStateEnum>(opState.OperationalStateID);
+
+    ReturnErrorOnFailure(mOperationalStateDataProvider.UseOpStateIDGetOpStateStruct(mEndpointId, mClusterId, op));
+
+    if (&mOperationalState != &opState)
+    {
+        char * dest = mOperationalState.OperationalStateLabel;
+        size_t len  = op.operationalStateLabel.size();
+        memset(dest, 0, sizeof(mOperationalState.OperationalStateLabel));
+        memcpy(dest, op.operationalStateLabel.data(), len);
+        mOperationalState.OperationalStateID = static_cast<uint8_t>(op.operationalStateID);
+    }
+    return CHIP_NO_ERROR;
+}
+
+const OperationalStateStruct & OperationalStateServer::GetOperationalState()
+{
+    return mOperationalState;
+}
+
+CHIP_ERROR OperationalStateServer::SetOperationalError(OperationalErrorStateStruct & op)
+{
+
+    mOperationalError.ErrorStateID = op.ErrorStateID;
+
+    memset(mOperationalError.ErrorStateLabel, 0, sizeof(mOperationalError.ErrorStateLabel));
+    memset(mOperationalError.ErrorStateDetails, 0, sizeof(mOperationalError.ErrorStateDetails));
+    memcpy(mOperationalError.ErrorStateLabel, op.ErrorStateLabel, sizeof(op.ErrorStateLabel));
+    memcpy(mOperationalError.ErrorStateDetails, op.ErrorStateDetails, sizeof(op.ErrorStateDetails));
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR OperationalStateServer::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    switch (aPath.mAttributeId)
+    {
+    case OperationalState::Attributes::OperationalStateList::Id: {
+        OperationalStateStructDynamicList * pOpList = nullptr;
+        size_t size                                 = 0;
+        GetOperationalStateList(&pOpList, size);
+        if (size == 0)
+        {
+            err = aEncoder.EncodeNull();
+        }
+        else
+        {
+            return aEncoder.EncodeList([&](const auto & encoder) -> CHIP_ERROR {
+                for (OperationalStateStructDynamicList * pHead = pOpList; pHead != nullptr; pHead = pHead->Next)
+                {
+                    ReturnErrorOnFailure(encoder.Encode(*pHead));
+                }
+                ReleaseOperationalStateList(pOpList);
+                pOpList = nullptr;
+                return CHIP_NO_ERROR;
+            });
+        }
+    }
+    break;
+
+    case OperationalState::Attributes::OperationalState::Id: {
+        OperationalState::Structs::OperationalStateStruct::Type opState;
+
+        opState.operationalStateID    = static_cast<OperationalState::OperationalStateEnum>(mOperationalState.OperationalStateID);
+        opState.operationalStateLabel = CharSpan::fromCharString(mOperationalState.OperationalStateLabel);
```

Comment:
Again, this is assuming null-termination, which is broken.

Reply:
Fix, don't use CharSpan::fromCharString.